packages/headless/src/features/insight-user-actions/insight-user-actions-preprocessing.ts

```diff
@@ -0,0 +1,260 @@
+import {
+  UserAction,
+  UserActionsState,
+  UserActionTimeline,
+  UserActionType,
+  UserSession,
+} from './insight-user-actions-state';
+
+const MSEC_IN_SECOND = 1000;
+const SECONDS_IN_MINUTE = 60;
+const MAX_MINUTES_IN_SESSION = 30;
+const MAX_MSEC_IN_SESSION =
+  MAX_MINUTES_IN_SESSION * SECONDS_IN_MINUTE * MSEC_IN_SECOND;
+
+interface rawUserAction {
+  name: string;
+  value: string;
+  time: string;
+}
+
+export const preprocessActionsData = (
+  state: UserActionsState,
+  actions: Array<rawUserAction>
+): UserActionTimeline => {
+  const ticketCreationDate = state.ticketCreationDate;
+  const excludedCustomActions = state.excludedCustomActions;
+
+  // If there is no ticket creation date or no actions, we return an empty timeline.
+  if (!ticketCreationDate || actions.length === 0) {
+    return {
+      precedingSessions: [],
+      session: undefined,
+      followingSessions: [],
+      caseSessionFound: false,
+    };
+  }
+
+  const sortedActions: rawUserAction[] = sortActions(actions);
+
+  const mappedUserActions = mapUserActions(sortedActions);
+
+  const timeline = splitActionsIntoTimelineSessions(
+    mappedUserActions,
+    ticketCreationDate
+  );
+
+  // Filter out custom actions to be excluded from the timeline
+  const filteredTimeline = filterTimelineActions(
+    timeline,
+    excludedCustomActions
+  );
+
+  return filteredTimeline;
+};
+
+// Sorting actions by most recent
+export const sortActions = (actions: rawUserAction[]) => {
+  const sortedActionsByMostRecent = (a: rawUserAction, b: rawUserAction) =>
+    Number(b.time) - Number(a.time);
+
+  return [...actions].sort(sortedActionsByMostRecent);
+};
+
+export const filterActions = (
+  actions: UserAction[],
+  excludedCustomActions: string[]
+): UserAction[] => {
+  const filteredActions = actions.filter((action) => {
+    let shouldExcludeCustomAction = false;
+    const eventType = action.eventData?.type || '';
+    const eventValue = action.eventData?.value || '';
+
+    if (eventType && eventValue) {
+      shouldExcludeCustomAction =
+        excludedCustomActions.includes(eventType) ||
+        excludedCustomActions.includes(eventValue);
+    }
+    return (
+      action.actionType !== UserActionType.CUSTOM || shouldExcludeCustomAction
+    );
+  });
+  return filteredActions;
+};
+
+export const mapUserActions = (rawActions: rawUserAction[]): UserAction[] => {
+  const mappedUserActions = rawActions.map((rawAction) => {
+    const actionData = JSON.parse(rawAction.value);
```

Comment:
you should be cautious here of the case where `actionData` will be undefined/null, it will cause an error in this case when you try then to access `actionData.event_type` for example.

Reply:
the `UserAction` interface already support potential undefined values. Also, I checked a lot of raw user events, couldn't find a case with undefined or null value, is there a situation where an event would have only a name and a time returned by the API?

Reply:
You can't trust the API. 
Yes you must assume that it's possible that maybe someday you'll receive an action with a `value` of `undefined` or a value that throws an error when you JSON.parse it.

I would put a try..catch around the JSON.parse, and just drop the action if I can't parse it.